flux-syntax/src/surface.rs

```diff
@@ -192,6 +194,8 @@ pub enum TyKind<R = ()> {
     Constr(Expr, Box<Ty<R>>),
     Tuple(Vec<Ty<R>>),
     Array(Box<Ty<R>>, ArrayLen),
+    /// The first `R` parameter is for the `DefId` corresponding to the hir OpaqueTy
+    Opaque(R, Option<Vec<GenericArg<R>>>, Bounds<R>),
```

Comment:
The `Option<Vec<GenericArg<R>>` looks sketchy and I think is wrong. If I understand correctly, the args in [`rustc_hir::TyKind::OpaqueDef`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/enum.TyKind.html#variant.OpaqueDef) correspond to the lifetimes captured by the opaque type so we should be generating a list of lifetime holes instead of type holes.

In any case, I wouldn't deal with this during resolving but during desugaring. So, I would eliminate the `Option<Vec<GenericArg<R>>>` field here, and during resolving we just get the `DefId`. Then during desugaring, we *query* the `Generics` of the opaque type and generate corresponding holes there.

This is also related to opaque types depending on refinement parameters.